Store: Add ability to add/edit BACS account info.

timmyc opened the pull request:
This closes #16630. Currently the BACS payment type is hidden on the settings screen due to not exposing the requisite fields to add bank account information. Now that we have added the ability to do so via the API via `wc-calypso-bridge`, this branch adds in the form fields to enable editing bank account data.

One 🔑  difference between this and `wp-admin` is we only allow the management of one bank account via the calypso interface. If the need arises, we could always revisit this choice in the future.

__Design__
I have just added in a new `<fieldset>` with a `<legend>` for the Account details - it looks a little rough at the moment, and I'm not sure we need to have the Account Details bit, but hopefully @jameskoster or @kellychoffman can chime in on that part:

![payment_settings_ _timmy_s_flies_ _wordpress_com](https://user-images.githubusercontent.com/22080/33226207-5a027946-d13d-11e7-90a9-18626f6fa8c3.png)

__To Test__
- Visit your store settings page http://calypso.localhost:3000/store/settings/low-value-long-named-domain.blog and note that the BACS option ( Bank Account ) is now shown
- Click manage, and fill in some form fields for account details and click Done
- Save the settings, hard refresh, and verify all your changes have persisted

jameskoster commented:
I think we should;

* Add a description after the title input to explain that this is how customers will see the payment method advertised
* Add a line to the "Instructions for customer at checkout" that explains that bank details for payment will be sent via email.
* Move the bank details fields above the instructions
* "Assemble" instructions for the email as they fill in the bank details. 

So, something like this;

![bacs](https://cldup.com/j2NHirHxp2.png)

kellychoffman commented:
Agree with Jay. 

Also: Are all these fields required? Do we do any sort of form validation? 

Does everyone know what IBAN and BIC / Swift is or should we add a `i` tooltip to those? (I have no idea what they mean, but I'm a dumb American)

allendav commented:
@jameskoster @kellychoffman - I wonder if we should have a password-like show/hide on the account number and routing number - i.e. to avoid sensitive information being displayed and subject to shoulder surfing

allendav commented:
@timmyc wrote:

> Does everyone know what IBAN and BIC / Swift is or should we add a i tooltip to those? (I have no idea what they mean, but I'm a dumb American)

If the store is US or CA based, no need to even include this prompt I think: https://www.iban.com/structure.html

allendav commented on client/extensions/woocommerce/app/settings/payments/payment-method-bacs.js:
```diff
@@ -0,0 +1,182 @@
+				} ),
+				accounts: PropTypes.shape( {
+					id: PropTypes.string.isRequired,
+					value: PropTypes.array.isRequired,
```
Might help flush out future bugs to proptype the value elements too

allendav commented on client/extensions/woocommerce/app/settings/payments/payment-method-bacs.js:
```diff
@@ -0,0 +1,182 @@
+			]
+		);
+
+		return accountData.length ? accountData[ 0 ] : {};
```
I take it that having multiple elements in the value array is impossible?

timmyc replied:
Correct in wp-admin you can add multiple accounts - but for simplicity sake in Calypso we are supporting one.

allendav replied:
Maybe enter that (supporting multiple) as a new low priority issue or a here as a TODO

allendav commented:
As I started entering an account name, I got a proptype warning

<img width="1429" alt="screen shot 2017-11-27 at 11 05 12 am" src="https://user-images.githubusercontent.com/1595739/33284161-f36bac84-d362-11e7-90cb-7bf09977bca7.png">

allendav commented:
Noteworthy: We might want to reword the prompt a little. The instructions are not just used in the email but on the order-received page too, at least for Storefront

<img width="1088" alt="payment-settings" src="https://user-images.githubusercontent.com/1595739/33284469-15dc7004-d364-11e7-986c-8437f6f87b11.png">

<img width="1429" alt="order-received" src="https://user-images.githubusercontent.com/1595739/33284484-1aec2f6c-d364-11e7-94ed-88198d597265.png">

allendav commented on client/extensions/woocommerce/app/settings/payments/payment-method-bacs.js:
```diff
@@ -0,0 +1,182 @@
+						/>
+					</FormFieldset>
+					<FormFieldset className="payments__method-edit-field-container">
+						<FormLabel>{ translate( 'BIC / Swift' ) }</FormLabel>
```
Maybe add context here and maybe for IBAN too (i.e. at least spell out the acronyms)

timmyc commented:
> Also: Are all these fields required? Do we do any sort of form validation?

There doesn't seem to be any validation on these fields in wp-admin @kellychoffman 

> If the store is US or CA based, no need to even include this prompt I think: 

Could a US-based store have an overseas account that supports IBAN? My inclination is to just show it like it is in core.

jameskoster commented:
> I wonder if we should have a password-like show/hide on the account number and routing number - i.e. to avoid (slightly) sensitive information being displayed and subject to shoulder surfing

I think that's a good idea. We have a password input component, let's use that.

> Could a US-based store have an overseas account that supports IBAN? My inclination is to just show it like it is in core.

That seems very unlikely to me?

I can't find a universal SWIFT/BIC lookup so all we can really do is recommend folks speak to their bank for this data. It's not required unless people want to accept international payments though. Perhaps we should mention that? Or even just hide the input if they don't ship internationally?

> Noteworthy: We might want to reword the prompt a little. The instructions are not just used in the email but on the order-received page too

Let's change that label from "Instructions for customer in order email notification" to "Instructions for customer upon order completion". That should cover all forms of order completion communication.

timmyc commented:
To keep things simple here, it _feels_ like we should not even show IBAN BIC/SWIFT with our user base being only US/CA currently. When the time comes to open up additional countries we can then revisit. Does that sound okay @jameskoster?

I'll apply the other changes requested above today.

jameskoster commented:
Well, there is the edge case of stores that;

* Ship internationally
* Want to offer BACS payments for international customers

Seems safe to assume that those store owners will be wanting payments sent to US/Canada banks though.

Would it be possible to check their shipping rules and only display the inputs for BIC/Swift if they ship internationally? Or to make things simpler, hide the inputs behind a "I want to offer BACS payments to international customers" checkbox?

justinshreve commented:
> Would it be possible to check their shipping rules and only display the inputs for BIC/Swift if they ship internationally? Or to make things simpler, hide the inputs behind a "I want to offer BACS payments to international customers" checkbox?

I think we should go the checkbox route right now, and add some tracking to see how often 1) BACS in general gets used and 2) How often a user uses the international fields.

The checkbox probably doesn't even need to save it's state anywhere. It can be auto-checked if those fields are already filled out, and otherwise the user can toggle it to show the fields.

timmyc commented:
Thanks again all for the input! In 8f0a300 I addressed the PropType issue, added more details to the PropTypes for account `shape`, added protected form inputs for Routing/Account number, and made the International bits toggled via a checkbox:

![bacs-updates](https://user-images.githubusercontent.com/22080/33353654-a7e32e66-d463-11e7-8fbf-85926c3689ac.gif)

jameskoster commented:
Lookin' good! I think we should change the checkbox label to;

> I want to offer BACS payment as an option to international customers

I think we should also highlight the required fields and add some descriptions;

Account name
This is for your reference

IBAN (and BIC / SWIFT)
You can obtain your IBAN number from your bank

timmyc commented:
> I think we should also highlight the required fields

Well as-is, nothing is a required field on the form right now. Much like wp-admin. I fear if we try to get too fancy with required fields in this form, we could miss some edge cases and make the form a bit of a chore to work with...

In fact I don't think any of the Payment Method forms have validation right now. Just checked PayPal you can submit the form with no email address.

I'll add in the labels but thinking we should punt on validation and maybe look at layering that in across payment types potentially.
